core/vm/eof.go

```diff
@@ -0,0 +1,373 @@
+			var arg int16
+			// Read immediate argument.
+			if err := binary.Read(bytes.NewReader(code[i+1:]), binary.BigEndian, &arg); err != nil {
+				return ErrEOF1InvalidRelativeOffset
+			}
+			// Calculate relative target.
+			pos := i + 3 + int(arg)
+
+			// Check if offset points to out-of-bounds location.
+			if pos < 0 || pos >= len(code) {
+				return ErrEOF1InvalidRelativeOffset
+			}
+			// Check if offset points to non-code segment.
+			// TODO(matt): include CALLF and RJUMPs in analysis.
+			if !analysis.codeSegment(uint64(pos)) {
+				return ErrEOF1InvalidRelativeOffset
+			}
+			i += 3
+		case opcode == CALLF:
+			var arg int16
+			// Read immediate argument.
+			if err := binary.Read(bytes.NewReader(code[i+1:]), binary.BigEndian, &arg); err != nil {
+				return fmt.Errorf("%v: %v", ErrEOF1InvalidCallfSection, err)
+			}
+			if int(arg) >= codeSections {
+				return fmt.Errorf("%v: want section %v, but only have %d sections", ErrEOF1InvalidCallfSection, arg, codeSections)
+			}
+			i += 3
+		default:
+			i++
+		}
+	}
+	if !opcode.isTerminating() {
+		return fmt.Errorf("%v: %v", ErrEOF1TerminatingInstructionMissing, opcode)
+	}
+	return nil
+}
+
+// uncheckedParseContainer parses an EOF v1 container without performing
+// validation.
+func uncheckedParseContainer(code []byte) *EOF1Container {
+	var c EOF1Container
+	// If the code has been validated (e.g. during deployment), skip the
+	// full validation.
+	c.typeSize, c.codeSize, c.dataSize = parseHeaderUnchecked(code)
+
+	// Set index to first byte after EOF header.
+	idx := c.HeaderSize()
+
+	// Read type section if it exists.
+	if typeSize := c.typeSize; typeSize != 0 {
+		c.types = parseTypeSection(code[idx : idx+int(typeSize)])
+		idx += int(typeSize)
+	}
+	// Calculate starting offset for each code section.
+	for _, size := range c.codeSize {
+		c.codeOffsets = append(c.codeOffsets, uint64(idx))
+		idx += int(size)
+	}
+	// Set data offset.
+	c.dataOffset = uint64(idx)
+	return &c
+}
+
+// parseHeaderUnchecked parses an EOF v1 header without performing validation.
+func parseHeaderUnchecked(code []byte) (typeSize uint16, codeSize []uint16, dataSize uint16) {
+	var (
+		i            = sectionHeaderStart
+		codeSections = uint16(1)
+	)
+	// Try to read type section.
+	if code[i] == byte(kindType) {
+		size, _ := parseSectionSize(code, i+1)
+		typeSize = size
+		codeSections = size / 2
+		i += 3
+	}
+	i += 1
+	// Read code sections.
+	for j := 0; j < int(codeSections); j++ {
+		size := binary.BigEndian.Uint16(code[i+3*j : i+3*j+2])
+		codeSize = append(codeSize, size)
+	}
+	i += int(2 * codeSections)
+	// Try to read data section.
+	if code[i] == byte(kindData) {
+		dataSize = binary.BigEndian.Uint16(code[i+1 : i+3])
+	}
+	return
+}
+
+// parseTypeSection parses an EOF type section.
+func parseTypeSection(code []byte) (out []Annotation) {
+	for i := 0; i < len(code); i += 2 {
+		sig := Annotation{
+			Input:  code[i],
+			Output: code[i+1],
```

Comment:
This is a crasher! The end-condition needs to be `len(code)-1`

```
fuzz: minimizing 72-byte failing input file
fuzz: elapsed: 42s, minimizing
--- FAIL: Fuzz (41.60s)
    --- FAIL: Fuzz (0.00s)
        testing.go:1356: panic: runtime error: index out of range [3] with length 3
            goroutine 286259 [running]:
            runtime/debug.Stack()
            	/usr/local/go/src/runtime/debug/stack.go:24 +0xdb
            testing.tRunner.func1()
            	/usr/local/go/src/testing/testing.go:1356 +0x1f2
            panic({0x7b4720, 0xc00a742528})
            	/usr/local/go/src/runtime/panic.go:884 +0x212
            github.com/ethereum/go-ethereum/core/vm.parseTypeSection(...)
            	/home/user/go/src/github.com/ethereum/go-ethereum/core/vm/eof.go:368
            github.com/ethereum/go-ethereum/core/vm.(*EOF1Container).UnmarshalBinary(0xc006e8e580, {0xc006c73750, 0xf, 0x10})
            	/home/user/go/src/github.com/ethereum/go-ethereum/core/vm/eof.go:112 +0x737
            github.com/ethereum/go-ethereum/core/vm.ParseEOF1Container(...)
            	/home/user/go/src/github.com/ethereum/go-ethereum/core/vm/eof.go:83
            github.com/ethereum/go-ethereum/core/vm.ParseAndValidateEOF1Container({0xc006c73750, 0xf, 0x10}, 0x4?)
```

Reply:
Nice catch!

Reply:
Okay good